haskell generic-builder: Use strictDeps always

Ericson2314 opened the pull request:
###### Motivation for this change

This helps avoid the `ARG_MAX` issues we've been having, and is generally a good idea to ensure cross comparability anyways.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
haskell generic-builder: Use strictDeps always

This helps avoid the `ARG_MAX` issues we've been having, and is
generally a good idea to ensure cross comparability anyways.

peti commented:
https://hydra.nixos.org/eval/1460999 is a build of the Haskell package set with this change applied.

IvanMalison commented:
@peti @Ericson2314 Is there anything blocking merge here?

peti commented:
Yes, the test builds are not complete yet.

IvanMalison commented:
@Ericson2314 It seems like strictDeps does not ameliorate this issue when it is encountered with stack builds (or maybe I'm doing something totally incorrectly). Do you have any idea how to fix the stack build of taffybar?

IvanMalison commented:
@Ericson2314 Actually, it seems that what happens when you enable crossConfig/stictDeps with the stack builder is that all of the haskell-gi generated bindings fail with 

```
setup: The program 'pkg-config' version >=0.9.0 is required but it could not
```

I suspect that this is probably because the paths that are passed to the stack build are not properly inherited in the builds for each haskell package, but I have to say that I'm pretty out of my depth with this haskell/stack/nix stuff.

peti reviewed (changes requested):
This change breaks `haskell.packages.ghc843.hspec-discover`. See the "new failing jobs" tab at https://hydra.nixos.org/eval/1460999 for further details.

Ericson2314 committed:
hspec{,-core,-discovery}: Add build-tool-depends to testsuite

`fetchpatch`ing a PR I just opened. This ought to get Cabal to make the
`PATH` include the newly-built exe, but unfortunately it doesn't.

Ericson2314 committed:
fetchpatch: Add includes to compliment excludes, and require that both not be non-empty.

Ericson2314 commented:
Weird! This is because GHC cannot find the testsuite from the same package to use a pre-processor. We have tests for this in Cabal https://github.com/haskell/cabal/tree/master/cabal-testsuite/PackageTests/BuildToolDepends; I'd know as I wrote them!

Ericson2314 commented:
https://hydra.nixos.org/eval/1464500

Ericson2314 commented:
Looks like there's some lingering issues with build tools: `hspec-discover` despite my patch, and maybe`alex` too. If anyone has any ideas I'd love to hear them.

Ericson2314 commented:
Rebased on https://github.com/NixOS/nixpkgs/pull/42247 which should work to keep things moving.

Ericson2314 commented:
https://hydra.nixos.org/eval/1465142. It will probably fail again, but I'd love to be wrong!

Ericson2314 commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1045,12 +1045,12 @@ self: super: {
-  taffybar = super.taffybar.overrideDerivation (drv: {
-    strictDeps = true;
+  # https://github.com/hspec/hspec/pull/355
+  hspec-discover_2_5_1 = appendPatch super.hspec-discover_2_5_1 (pkgs.fetchpatch {
```
OK I'll try

Ericson2314 replied:
Ah ok fixed. didn't realize this was the executable from `hspec-meta`, not `hspec-discover`.

Ericson2314 commented:
OK fixed this build, reevaluating. We'll see what fails.

Ericson2314 commented:
Huh. My evaluation hydra evaluation is stuck pending, and just disappeared before that.

Ericson2314 commented:
https://hydra.nixos.org/eval/1465755 finally getting somewhere I hope.

Ericson2314 committed:
haskell generic-builder: Make test and benchmark tool depends `nativeBuildInputs`

I suppose I forgot these when I did this before.

Ericson2314 committed:
haskell lib: Make helpers to add benchmark and test tool depends

These seems to be common omissions due to a) non-new-build being overly
flexible, and b) code gen most often used for test boilerplate.

Ericson2314 committed:
base-compat: Add missing build tool dependency

Ericson2314 committed:
alex: `which` needs to be a tool dep so it's a nativeBuildInput

Also pull it from `buildPackages`, as that is where nativeBuildInputs
come from.

matthewbauer committed:
with-location: add hspec-discover

matthewbauer committed:
text-conversions: add hspec-discover

matthewbauer committed:
haskell: add more test tool depend for hspec-discover

Needed for new ‘strictDeps = true;’ handling.

Including-

- slim
- string-conversions
- skeletal-test
- catamorphism
- unliftio
- logging-facade
- distributive
- doctest
- http-types
- interpolate
- mockery
- mime-mail
- http2

Ericson2314 committed:
haskell with-location: Link upstreaming attempt

Ericson2314 committed:
with-location: Link upstreamming attempt

Ericson2314 commented:
@23Skidoo BTW this forces the use of `build-tool-depends` (or it's cabal2nix equivalents via hacks) for packages to get binaries. How strict ought we to be / what is the comparability story planned for Cabal 3?

matthewbauer commented:
How much breakage are we allowed to have? Packages *should* handle this themselves and they will also get this error in cabal new-build. We are now below 500 failures. Is this okay with everyone or should we keep adding more?

Ericson2314 commented on pkgs/build-support/fetchpatch/default.nix:
```diff
@@ -5,7 +5,7 @@
 
 { lib, fetchurl, patchutils }:
-{ stripLen ? 0, extraPrefix ? null, excludes ? [], ... }@args:
+{ stripLen ? 0, extraPrefix ? null, excludes ? [], includes ? [], ... }@args:
```
@matthewbauer Yeah but it's a false positive. Cause it's a fixed-output derivation there isn't any rebuild.

matthewbauer committed:
haskell-modules: Fix mockery in macos

Weird errors otherwise - seems to need this to know it can skip the
bad test.

matthewbauer committed:
haskell-modules: Add many more overrides for hspec-discover

This is needed for picking up hspec-discover in strictDeps generic-builder.nix.

matthewbauer committed:
haskell-modules: Add test tool depend for modern-uri

matthewbauer committed:
haskell-modules: Add pandoc for heist

matthewbauer committed:
haskell-modules: Add more test tool depends

domenkozar commented:
@angerman has a patch to fix the ARG_MAX issue https://usercontent.irccloud-cdn.com/file/VMw0fTI2/0001-Stop-haskell-from-generating-out-lib.patch

matthewbauer committed:
haskell-modules: Add more overrides

domenkozar commented:
@matthewbauer could you rebase on top of master?

matthewbauer committed:
yesod-persistent: add hspec-discover for 1.4.3

domenkozar committed:
configuration-common.nix: fix build tools

domenkozar commented:
This change requires a lot of manual overrides. I think solution @angerman is proposing is less intrusive. Another way is to actually fix the commit that made MAX_ARGS explode: https://github.com/NixOS/nixpkgs/pull/41939/files#diff-360bed088faac7580cee137eeffb39e0R279

Ericson2314 commented:
@domenkozar the big change motivating all the overrides here is we're being explicit for the first time about when we need a binary from another package or just its library. So I think these changes will be very good for multiple outputs (https://github.com/NixOS/nixpkgs/pull/32629). Also, Cabal 3.0 due at the end of the year will remove the non-new-build stuff, in which case being explicit in the Cabal file will also be required.

My guess is Cabal will need to come up with some sort of back-compat story, and we can replace most the overrides with that, probably implemented in Cabal2nix but who knows.

domenkozar commented:
Oh you mean stuff like this: https://github.com/sol/with-location/pull/1

OK, but why is this needed in order to unbreak ARG_MAX? :)

domenkozar commented:
Could someone pull in https://github.com/domenkozar/nixpkgs/commit/4149778f478e5da33c9bffa2b1d8e5a81cafd253

matthewbauer committed:
configuration-ghc-8.4.x.nix: Update hspec versions

These have changed on master.

domenkozar commented:
@matthewbauer btw adding build tool dependencies should go to `configuration-nix` as it's not version dependent :) Thanks!

domenkozar commented:
FYI: it was reverted and I agree with @peti we can autogenerate these: https://github.com/NixOS/cabal2nix/issues/356
